### PR TITLE
Upgrade CI with Rubocop

### PR DESCRIPTION
**Pull request by jterapin**
*Description of changes:*
- Updated CI to include rubocop task for codegen directory
- Added rubocop yml file
- Fixed rubocop offenses in codegen

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by alextwoods — COMMENTED** [review]
Nice!

File: codegen/.rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+Layout/LineLength:
+  Max: 80
+
+Metrics/MethodLength:
+  Max: 20
+
+Style/Documentation:
+  Exclude:
+    - '**/integration-specs/*_spec.rb'
+    - 'smithy-ruby-codegen-test-utils/**/*'
```
[the remark this answers is not in the document]
Comment (alextwoods) [reply codegen/.rubocop.yml:25]: Sorry, to clarify - I think just the middelware/plugins that are int he codegen-test-utils should have documentation, not all specs!

File: .github/workflows/codegen_ci.yml [hunk 2]
```diff
@@ -117,3 +117,29 @@ jobs:
...
+  codegen-rubocop:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Setup Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 3.2
+
+      - name: Build and install Hearth
```
Comment (alextwoods) [inline .github/workflows/codegen_ci.yml:130]: I think you can remove this - we shouldn't need hearth installed to run rubocop.

Comment (jterapin) [reply .github/workflows/codegen_ci.yml:130]: Gotcha, I will remove now.

File: codegen/.rubocop.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+AllCops:
```
Comment (mullermp) [inline codegen/.rubocop.yml:1]: Is there a way we can use hearth's rubocop yml so we don't have a duplicate?

File: .github/workflows/codegen_ci.yml [hunk 4]
```diff
@@ -117,3 +117,25 @@ jobs:
...
+    steps:
+      - name: Setup Ruby
+        uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 3.2
+
+      - uses: actions/checkout@v2
+
+      - name: Copy Gemfile and dependencies
+        run: cp hearth/Gemfile hearth/hearth.gemspec hearth/VERSION codegen/
```
Comment (mullermp) [inline .github/workflows/codegen_ci.yml:133]: Why do we need gemspec and version?

Comment (jterapin) [reply .github/workflows/codegen_ci.yml]: Gemfile needs both to run the bundle install per our convo

File: codegen/.rubocop.yml [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    - 'projections/**/*'
+    - '**/build/**/*'
+
+Metrics:
+  Exclude:
+    - '**/**/*_spec.rb'
+
+Metrics/BlockLength:
+  Exclude:
+    - '**/**/*_spec.rb'
```
Comment (mullermp) [inline codegen/.rubocop.yml:14]: Double star is only needed once, it catches everything

**Comment by jterapin** [comment]
Based on our group discussion yesterday, I made the following changes:
- Codegen's rubocop file will inherit base configuration files from hearth's `.rubocop.yml` via using the `inherit_from`
- Updated Hearth's `.rubocop.yml` file relating to `Gemspec/MFA` so it can be inherited
- Still bringing over Hearth's Gemfile (and its dependencies) to run the rubocop in codegen (ensuring consistency in rubocop versions if we decide to specify)

Thanks for looking this throughly, team!

**Review by alextwoods — APPROVED** [review]
nice!

**Review by mullermp — APPROVED** [review]

**Pull request merged**